test/imath/bind/imath_vec.cpp

```diff
@@ -50,13 +57,29 @@ template <class T> class Vec3 {
     constexpr static T baseTypeSmallest();
     constexpr static T baseTypeEpsilon();
 
+    template <class S> IMATH_HOSTDEVICE void setValue (S a, S b, S c);
+    */
+
 } CPPMM_VALUETYPE;
 
 // These type aliases will cause the class template to be monomorphized with
 // the given template parameter as if it had been explicitly declared
 using V3f = Vec3<float>;
+V3f v3f;
 using V3i = Vec3<int>;
-
-} // namespace IMATH_INTERNAL_NAMESPACE
+V3i v3i;
```

Comment:
I think it's the variable declaration that causes the template to be monomorphized.

Making a global will work for types that are simple to construct, but might be more difficult for types that rely on external state for construction, parameters in the constructor (although I cant think of any concrete examples off the top of my head). Rather than implicitly monomophizing the template through a variable declaration of an alias, would it not be better to explicitly instantiate the templates directly?

I know this is only an example.. so not a big deal.

Reply:
What exactly do you mean by "explicitly insantiate the templates directly"?

Reply:
Sorry really bad spelling. I meant 'explicitly instantiate the templates'
https://stackoverflow.com/questions/2351148/explicit-template-instantiation-when-is-it-used

Forcing the template expansion, like how you do when you want the template implementation to be pre-compiled for a certain type.
```
template class Vec3<int>;
using V3i = Vec3<int>;
```

I've noticed that with clang. When you do that, incomplete types become complete and become navigable.

Reply:
Ahhhh thanks. I hadn’t really registered that that was a thing. That’s the beauty of C++ - there’s always new features to learn about!

That does seem like exactly what we need!

Reply:
So on further investigation we can remove the construction of the type by using explicit template instantiation. but only if we instantiate the template for both the binding type and the library type

Reply:
bugger that's annoying.